runtime/gc_vlhgc/CopyForwardScheme.cpp

```diff
@@ -4073,21 +4076,45 @@ class MM_CopyForwardSchemeRootClearer : public MM_RootScanner
 #if defined(J9VM_GC_ENABLE_DOUBLE_MAP)
 	virtual void doDoubleMappedObjectSlot(J9Object *objectPtr, struct J9PortVmemIdentifier *identifier) {
 		MM_EnvironmentVLHGC *env = MM_EnvironmentVLHGC::getEnvironment(_env);
-		env->_copyForwardStats._doubleMappedArrayletsCandidates += 1;
+		env->_copyForwardStats._offHeapRegionCandidates += 1;
 		if (!_copyForwardScheme->isLiveObject(objectPtr)) {
 			Assert_MM_true(_copyForwardScheme->isObjectInEvacuateMemory(objectPtr));
 			MM_ForwardedHeader forwardedHeader(objectPtr, _extensions->compressObjectReferences());
 			objectPtr = forwardedHeader.getForwardedObject();
 			if (NULL == objectPtr) {
 				Assert_MM_mustBeClass(_extensions->objectModel.getPreservedClass(&forwardedHeader));
-				env->_copyForwardStats._doubleMappedArrayletsCleared += 1;
+				env->_copyForwardStats._offHeapRegionsCleared += 1;
 				OMRPORT_ACCESS_FROM_OMRVM(_omrVM);
 				omrvmem_release_double_mapped_region(identifier->address, identifier->size, identifier);
 			}
 		}
 	}
 #endif /* J9VM_GC_ENABLE_DOUBLE_MAP */
 
+#if defined(J9VM_GC_ENABLE_SPARSE_HEAP_ALLOCATION)
+	virtual void doObjectInVirtualLargeObjectHeap(J9Object *objectPtr) {
+		MM_EnvironmentVLHGC *env = MM_EnvironmentVLHGC::getEnvironment(_env);
+		env->_copyForwardStats._offHeapRegionCandidates += 1;
+
+		if (!_copyForwardScheme->isLiveObject(objectPtr)) {
+			Assert_MM_true(_copyForwardScheme->isObjectInEvacuateMemory(objectPtr));
+			void *dataAddr = _extensions->indexableObjectModel.getDataAddrForContiguous((J9IndexableObject *)objectPtr);
+			MM_ForwardedHeader forwardedHeader(objectPtr, _extensions->compressObjectReferences());
+			objectPtr = forwardedHeader.getForwardedObject();
+			/* If forwarded object is NULL, free the sparse region occupied by the data of the indexable object */
+			if (NULL == objectPtr) {
+				Assert_MM_mustBeClass(_extensions->objectModel.getPreservedClass(&forwardedHeader));
+				env->_copyForwardStats._offHeapRegionsCleared += 1;
+				_extensions->largeObjectVirtualMemory->freeSparseRegionAndUnmapFromHeapObject(_env, dataAddr);
```

Comment:
seems like we will end up calling this multiple times for a single object. probalby it's a no-op (since we remove the mapping first time), but it's still suboptimal from perspective of this method (wasting time), and also stats candidates/clear will be inflated

Reply:
seems to be resolved by checking sparseHeapAllocation before entering this method